datumaro/components/progress_reporting.py

```diff
@@ -0,0 +1,93 @@
+# Copyright (C) 2022 Intel Corporation
+#
+# SPDX-License-Identifier: MIT
+
+from __future__ import annotations
+
+from typing import Iterable, Optional, Tuple, TypeVar
+import math
+
+T = TypeVar('T')
+
+class ProgressReporter:
+    def get_frequency(self) -> float:
+        """
+        Returns reporting frequency.
+
+        For example, 0.1 would mean every 10%.
+        """
+        raise NotImplementedError
+
+    def start(self, total: int, *, desc: Optional[str] = None):
+        """Initializes a progress bar"""
+        raise NotImplementedError
+
+    def report_status(self, progress: int):
+        """Updates a progress bar"""
+        raise NotImplementedError
+
+    def finish(self):
+        """Closes a progress bar"""
+        raise NotImplementedError
+
+    def iter(self, iterable: Iterable[T], *,
+            total: Optional[int] = None,
+            desc: Optional[str] = None
+    ) -> Iterable[T]:
+        """
+        Traverses the iterable and reports progress simultaneously.
+
+        Starts and finishes the progress bar automatically.
+
+        Args:
+            iterable: An iterable to be traversed
+            total: The expected number of iterations. If not provided, will
+              try to use iterable.__len__.
+            desc: The status message
+
+        Returns:
+            An iterable over elements of the input sequence
+        """
+
+        if total is None and hasattr(iterable, '__len__'):
+            total = len(iterable)
+
+        self.start(total, desc=desc)
+
+        try:
+            if total:
+                display_step = math.ceil(total * self.get_frequency())
+
+            for i, elem in enumerate(iterable):
+                if not total or i % display_step == 0:
+                    self.report_status(i)
+
+                yield elem
+        finally:
+            self.finish()
```

Comment:
I think the call to `finish` should just follow the loop and not be in a `finally` block, for a couple reasons:

1. If the loop terminates with an exception, then the task did _not_ finish. The client and/or the reporter can use the lack of a `finish` call to show that the task was not completed.
2. Even ignoring the above, this is unreliable. If an exception is raised in the caller of `iter` during iteration, then the `finally` block will not be executed until the generator is garbage-collected, and it's undefined when that happens.

Reply:
Fixed. I've added `close` and CM semantics.

Reply:
Why does it need CM semantics, though? If any progress bars are not completed, then the client can deal with that. Requiring services to close their own progress bars just serves to complicate their code.

Reply:
You have to destroy threads, because otherwise they'll be leaked. But nothing has to get leaked here. The client has a reference to the progress reporter, so it can clean it up, and the progress reporter can keep track of its children and clean them up too.

There are a lot more services (extractors, converters) than there are clients, so the burden of dealing with resources should fall on the latter. That way, the added complexity of services can be minimized.